### ISSUE TITLE
TinkerPop tests are failing in master branch (closed after 6 hours)

### ISSUE THREAD

**Issue by porunov** [JanusGraph/janusgraph#3747]
On February 22 we had the last time TinkerPop tests CI executed without any problem (scheduled weekly job).
Starting from March 1 the scheduled job failed and it's failing from that time each week. There were no any successful runs after that.
It either means: We made a bug somewhere from February 22 to March 1 or something changed in the CI execution (less powerful machines or something else. 
Previously TP tests job were taking 4:45 hours in average. From March 1 it takes over 6 hours, but GitHub actions kill TP tests after 6 hours.

Scheduled execution jobs can be found here: https://github.com/JanusGraph/janusgraph/actions?query=branch%3Amaster+event%3Aschedule

The problem was noticed in the next PRs: #3418 , #3743 

Link to the branch with 2 canceled after 6 hours recent jobs: https://github.com/porunov/janusgraph/actions/runs/4809004095/jobs/8559714661

- Version: master
- Storage Backend: cassandra (byteordered)
- Mixed Index Backend: none
- Expected Behavior: TP tests should pass
- Current Behavior: CI for TP tests takes 6 hours after that it's force closed by GitHub
- Steps to Reproduce:

For Cassandra 3:
```
mvn clean install --projects janusgraph-cql -DskipTests=true --batch-mode --also-make -Pjava-11
mvn verify --projects janusgraph-cql -Pjava-11 -Pcassandra3-byteordered -Dtest.skip.tp=false -DskipTests=true
```

For Cassandra 4:
```
mvn clean install --projects janusgraph-cql -DskipTests=true --batch-mode --also-make -Pjava-11
mvn verify --projects janusgraph-cql -Pjava-11 -Pcassandra4-byteordered -Dtest.skip.tp=false -DskipTests=true
```

### Stack Trace (if you have one)

```
Error: The operation was canceled.
```

**Comment by porunov** [comment]
Here is a list of commits which we had from February 22 to March 1 (the issue is most likely in one of those commits).
Commits sorted by date (starting from oldest commits and ending the most recent commits):
https://github.com/JanusGraph/janusgraph/commit/94b3da5bbb877cfc99388069700bc6fb224b1063
https://github.com/JanusGraph/janusgraph/commit/20e14a7a3998355db06c2bb2ceac57b7657de07b
https://github.com/JanusGraph/janusgraph/commit/69e01dba36d9c2512e108b9c614fa6e92892b848
https://github.com/JanusGraph/janusgraph/commit/96844943967eac40a0c5576d8db0969cba2d78cc
https://github.com/JanusGraph/janusgraph/commit/6725fd3114f81fe271ab4316500ae0b25ef9152f
https://github.com/JanusGraph/janusgraph/commit/3429215d89c064e200a79329be299816e5c69d76
https://github.com/JanusGraph/janusgraph/commit/245e79fcb10062a1d64d1d97ce6ec42ad434efa5
https://github.com/JanusGraph/janusgraph/commit/fd3b92c89634146e0d44831d9f1b0d0e61e98eb4
https://github.com/JanusGraph/janusgraph/commit/a44647a834faa47b50773ce7bdada328a2c3ceb2
https://github.com/JanusGraph/janusgraph/commit/ba672b9c281e22597ca930e03f67191c1db00ce2
https://github.com/JanusGraph/janusgraph/commit/49c5f8039ab6ac233e8361786e9a0571d9a11b99
https://github.com/JanusGraph/janusgraph/commit/cdb6a93ffddb5b26532d0ed91d6a2480dfe9e0ae
https://github.com/JanusGraph/janusgraph/commit/19a8637b8b787f3f62898873cd7372e107cccace
https://github.com/JanusGraph/janusgraph/commit/76f9a6b6a01881d39764b0af17b800162d7e17fb
https://github.com/JanusGraph/janusgraph/commit/671672b9cf13f02c982c92861128f37ec1d16bb6
https://github.com/JanusGraph/janusgraph/commit/fcc8b33298e7f00f5f39ba2ddd120d01cb2eba00
https://github.com/JanusGraph/janusgraph/commit/44755074e7733f41123ac5581df17a28b3185431
https://github.com/JanusGraph/janusgraph/commit/2c7007653a74ba01b938c7803a17cf17b1e3a9fe

**Comment by porunov** [comment]
> The last commit for which TP tests supped to passed on February 22 was made on February 17: [d30e789](https://github.com/JanusGraph/janusgraph/commit/d30e789e2468b359dad7fba9c1bbe2ca1479da86) (started TP tests to verify TP tests still pass for that commit [here](https://github.com/porunov/janusgraph/commit/e98d168109bfeaf3701794eb1b2facd77a5ff29e) . Branch is [here](https://github.com/porunov/janusgraph/tree/tp-tests/d30e789e2468b359dad7fba9c1bbe2ca1479da86))
> 
> Here is a list of commits which we had from February 22 to March 1 (the issue is most likely in one of those commits). Commits sorted by date (starting from oldest commits and ending the most recent commits):
> 
> [94b3da5](https://github.com/JanusGraph/janusgraph/commit/94b3da5bbb877cfc99388069700bc6fb224b1063) (started TP tests [here](https://github.com/porunov/janusgraph/commit/e332105e03cab93b60e81d60aec2bd348db95c63) . Branch is [here](https://github.com/porunov/janusgraph/tree/tp-test/94b3da5))
> 
> [20e14a7](https://github.com/JanusGraph/janusgraph/commit/20e14a7a3998355db06c2bb2ceac57b7657de07b) (started TP tests [here](https://github.com/porunov/janusgraph/commit/c99681075386b9bf4e077e7c66d4fa6ae07d4898) . Branch is [here](https://github.com/porunov/janusgraph/tree/tp-tests/20e14a7a3998355db06c2bb2ceac57b7657de07b))
> 
> [69e01db](https://github.com/JanusGraph/janusgraph/commit/69e01dba36d9c2512e108b9c614fa6e92892b848) (skipping checks for this commit due to doc only updates)
> 
> [9684494](https://github.com/JanusGraph/janusgraph/commit/96844943967eac40a0c5576d8db0969cba2d78cc) (Only configuration option description doc was changed here. Thus, skipping this commit)
> 
> [6725fd3](https://github.com/JanusGraph/janusgraph/commit/6725fd3114f81fe271ab4316500ae0b25ef9152f) (started TP tests [here](https://github.com/porunov/janusgraph/commit/d0646e790b7acebab6f44f16da53faccd6da69c5) . Branch is [here](https://github.com/porunov/janusgraph/tree/tp-tests/6725fd3114f81fe271ab4316500ae0b25ef9152f))
> 
> [3429215](https://github.com/JanusGraph/janusgraph/commit/3429215d89c064e200a79329be299816e5c69d76) (started TP tests [here](https://github.com/porunov/janusgraph/commit/b2bf578fe62c49a6e9ce54bd4ccaf4248b2f5a0b) . Branch is [here](https://github.com/porunov/janusgraph/tree/tp-tests/3429215d89c064e200a79329be299816e5c69d76))
> 
> [245e79f](https://github.com/JanusGraph/janusgraph/commit/245e79fcb10062a1d64d1d97ce6ec42ad434efa5)
> 
> [fd3b92c](https://github.com/JanusGraph/janusgraph/commit/fd3b92c89634146e0d44831d9f1b0d0e61e98eb4)
> 
> [a44647a](https://github.com/JanusGraph/janusgraph/commit/a44647a834faa47b50773ce7bdada328a2c3ceb2)
> 
> [ba672b9](https://github.com/JanusGraph/janusgraph/commit/ba672b9c281e22597ca930e03f67191c1db00ce2) (release candidate for which we already executed TP tests successfully. Thus, skipping this commit checks)
> 
> [49c5f80](https://github.com/JanusGraph/janusgraph/commit/49c5f8039ab6ac233e8361786e9a0571d9a11b99) (snapshot return commit. Skipping this commit)
> 
> [cdb6a93](https://github.com/JanusGraph/janusgraph/commit/cdb6a93ffddb5b26532d0ed91d6a2480dfe9e0ae) (doc only change. Skipping this commit)
> 
> [19a8637](https://github.com/JanusGraph/janusgraph/commit/19a8637b8b787f3f62898873cd7372e107cccace)
> 
> [76f9a6b](https://github.com/JanusGraph/janusgraph/commit/76f9a6b6a01881d39764b0af17b800162d7e17fb)
> 
> [671672b](https://github.com/JanusGraph/janusgraph/commit/671672b9cf13f02c982c92861128f37ec1d16bb6)
> 
> [fcc8b33](https://github.com/JanusGraph/janusgraph/commit/fcc8b33298e7f00f5f39ba2ddd120d01cb2eba00)
> 
> [4475507](https://github.com/JanusGraph/janusgraph/commit/44755074e7733f41123ac5581df17a28b3185431)
> 
> [2c70076](https://github.com/JanusGraph/janusgraph/commit/2c7007653a74ba01b938c7803a17cf17b1e3a9fe) (doc only change. Skipping this commit)

Based on the above investigation as the last successful TP tests execution were on February 26 ([this commit](https://github.com/JanusGraph/janusgraph/commit/ba672b9c281e22597ca930e03f67191c1db00ce2)) the scope of potential harmful commits is reduced to the above list (unless the issue is in GitHub actions).

https://github.com/JanusGraph/janusgraph/commit/19a8637b8b787f3f62898873cd7372e107cccace

https://github.com/JanusGraph/janusgraph/commit/76f9a6b6a01881d39764b0af17b800162d7e17fb

https://github.com/JanusGraph/janusgraph/commit/671672b9cf13f02c982c92861128f37ec1d16bb6

https://github.com/JanusGraph/janusgraph/commit/fcc8b33298e7f00f5f39ba2ddd120d01cb2eba00

https://github.com/JanusGraph/janusgraph/commit/44755074e7733f41123ac5581df17a28b3185431

**Comment by porunov** [comment]
It’s strange but based on the above research the trouble commit was https://github.com/JanusGraph/janusgraph/commit/76f9a6b6a01881d39764b0af17b800162d7e17fb

Before that commit TP tests are passing (checked 3 different commits before that one).
After that commit TP tests are failing due to timeout issues (this trouble commit and the next commit failed. 2 more commits are still running but I suspect they should fail after they reach timeout).

**Comment by porunov** [comment]
> It’s strange but based on the above research the trouble commit was [76f9a6b](https://github.com/JanusGraph/janusgraph/commit/76f9a6b6a01881d39764b0af17b800162d7e17fb)
> 
> Before that commit TP tests are passing (checked 3 different commits before that one). After that commit TP tests are failing due to timeout issues (4 commits after that one is tested).

I tried to revert the failing commit here #3749 but looks like it didn't help.
Moreover @li-boxuan commented [here](https://github.com/JanusGraph/janusgraph/pull/3418#issuecomment-1524330899) that TP passed for some commits in #3418 .

It could be that above investigation of a wrong commit could be influenced by instability of GitHub Actions. That said, it could be that we just have now multiple commits in `master` branch which prevent TinkerPop tests from finishing in time. It is strange that before (and including) commit https://github.com/JanusGraph/janusgraph/commit/19a8637b8b787f3f62898873cd7372e107cccace TP tests are passing for multiple tries (as seen from [the above investigation](https://github.com/JanusGraph/janusgraph/issues/3747#issuecomment-1523953108)), but right after that commit TinkerPop are constantly failing both for scheduled TP tests on `master` and in the above investigation.

Thus, I think the problem is one of the below problems:
1. GitHub Actions are suddenly become unstable (maybe they reduced resources for GitHub Actions or did something else).
2.  The problem is in this commit https://github.com/JanusGraph/janusgraph/commit/76f9a6b6a01881d39764b0af17b800162d7e17fb and some other commits after that one.

It's a little bit tricky to find our the root cause of this issue and why TinkerPop tests are not passing for `cassandra3-byteordered` and `cassandra4-byteordered` in time anymore.

@JanusGraph/committers in case anyone can investigate this issue as well or have some thoughts about how to fix TP tests for GitHub Actions that would be great.
The problem is that we can't make new JanusGraph releases without TP tests passing. Finding a troubling commit is more challenging due to many commits after the last successful commit with TP tests and also due to the big time to test each commit.

**Comment by porunov** [comment]
More tests:

Finished:

1. Revert `checker-qual` to `3.30.0`. (**FAILED**). PR: #3749
2. Revert all dependencies but leave other changes. (**FAILED**). PR: #3752

Waiting for the results:

- Reset branch to https://github.com/JanusGraph/janusgraph/commit/cc0843e218c4cdc0e4ef7135a40bc04683d7f74f and revert all dependencies. Branch https://github.com/porunov/janusgraph/tree/bugfix/tp-tests-2 . Commit https://github.com/porunov/janusgraph/commit/bcfd80f76b30c0b8ee90f3a832e29c0ec265af2f .

**Comment by porunov** [comment]
Executing more tests shows that the last `rc-2` release had the same problem, but we were simply lucky that those tests passed. Re-running tp-tests for that commit - failed.
[Multiple re-runs of previous commits](https://github.com/JanusGraph/janusgraph/issues/3747#issuecomment-1523953108) showed that most likely the problem is in one of the followings commits:

https://github.com/JanusGraph/janusgraph/commit/245e79fcb10062a1d64d1d97ce6ec42ad434efa5 ~~(This commit was earlier than the latest successful TP tests execution commit. Thus skipping this commit)~~  (started TP tests [here](https://github.com/porunov/janusgraph/commit/ba87e6ab69024d70c06f278d4ac5654c4e2a8b9f) . Branch is [here](https://github.com/porunov/janusgraph/tree/bugfix/tp-tests-22))

https://github.com/JanusGraph/janusgraph/commit/fd3b92c89634146e0d44831d9f1b0d0e61e98eb4 ~~(This commit was earlier than the latest successful TP tests execution commit. Thus skipping this commit)~~  (started TP tests [here](https://github.com/porunov/janusgraph/commit/7e238b2246cd0b7fc119e997ddfb9e79e15361c9) . Branch is [here](https://github.com/porunov/janusgraph/tree/bugfix/tp-tests-21))

https://github.com/JanusGraph/janusgraph/commit/a44647a834faa47b50773ce7bdada328a2c3ceb2 ~~(This commit was earlier than the latest successful TP tests execution commit. Thus skipping this commit)~~  (started TP tests [here](https://github.com/porunov/janusgraph/commit/c4308b1735bc5db7a9b73e41e03b96db828f5e54) . Branch is [here](https://github.com/porunov/janusgraph/tree/bugfix/tp-tests-20))

Started tests executions for these commits to see which one may fail.
Notice, that sometimes job succeeds, thus if the job seceded only once it doesn't mean it will have the same behavior on a second run. We should test commit multiple times to see if it fails or not. If it fails at least once - there is a problem with the code under that commit (most likely).

**Comment by porunov** [comment]
**Update**: 

With the ongoing investigation the scope of the trouble commits is reduced to this commit only: https://github.com/JanusGraph/janusgraph/commit/a44647a834faa47b50773ce7bdada328a2c3ceb2
It also explains why only TinkerPop tests for `cassandra` backend were failing.

I noticed that both `byteordered` and `murmur` partitioning was failing. Thus, it's probably not related to particular partitioning. That said, I also noticed that the tests hang only for Java 11 and not for Java 8 (could be just coincidence).
Looks like with disabling custom CQL executor pool, some tinkerpop tests just hang sometimes. 
That said, the behavior isn't consistent. Sometimes tests are finished in 4:10 hours and sometimes they hang for more than 6 hours and then killed by GitHub Actions. 

The strange thing is that those TinkerPop CQL tests are either finishing in the next window [**from 4 hours to 4:50 hours**] or don't finish at all. From many times over these days when I was running tests I have never seen a time when those tests finish in 5:10, 5:30, 5:55 (i.e. they always finish under 5 hours). This brings me to thoughts that potentially it could be unrelated to CQL requests throughput but to something else (just a theory). 

In case anyone knows why would some TP tests hang for over 6 hours for CQL storage backend when CQL executor pool is disabled it would be really great if you could share your thoughts here because currently I'm not getting it. 
From my perspective the internal JanusGraph executor pool to handle CQL requests was just unnecessary overhead which would artificially limit the amount of parallel CQL requests (even so we have a separate parameter for that `storage.cql.max-requests-per-connection`). Thus, the usage of CQL executor pool is unnecessary and it adds just additional overhead in term of threads. 
It looks like I'm wrong about CQL executor pool and maybe it should be enabled, but I don't understand why.

The merged PR of that commit is #3607 

cc @li-boxuan @cdegroc because we had some related discussions here #3170

**Comment by porunov** [comment]
Started TP tests for `master` branch with enabled CQL executor pool: https://github.com/porunov/janusgraph/commit/9d346b8dc71ba4ff9228212f257b69f972faf983 (Branch https://github.com/porunov/janusgraph/tree/bugfix/tp-tests-30)

**Comment by porunov** [comment]
> Started TP tests for `master` branch with enabled CQL executor pool: [porunov@9d346b8](https://github.com/porunov/janusgraph/commit/9d346b8dc71ba4ff9228212f257b69f972faf983) (Branch https://github.com/porunov/janusgraph/tree/bugfix/tp-tests-30) (**PASSED 2 times**)
> 
> TP tests for #3760 but with executor service used for mutation queries (don't understand why would we use it for those queries because mutation queries don't have any deserialization work to be done as slice queries. Nevertheless, we do have 2 implementations for some reason (with and without executor service used). [porunov@f4d4b0e](https://github.com/porunov/janusgraph/commit/f4d4b0ed9688741b75b765b815ab12821194ea1e) (Branch https://github.com/porunov/janusgraph/tree/feature/cql-multi-key-slices-mutate-many-with-executor) (**PASSED 1 time**)
> 
> TP tests in #3760 before ensuring `PermanentBackendException` is thrown on `InterruptedException` (**FAILED 2 times**) TP tests in #3760 after ensuring `PermanentBackendException` is thrown on `InterruptedException` (**PASSED 2 times**)

As the above research showed. The long running TinkerPop tests were not related to the `ExecutorService` usage or performance. Instead, it was related to the bug in CQL storage backend which were propagating `InterruptedException` as `TemporaryBackendException` instead of `PermanentBackendException`. 
To cancel the query execution we should throw `PermanentBackendException` on `InterruptedException`.
This will be fixed as part of #3760
